addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/model/HttpChannelConfig.java

```diff
@@ -0,0 +1,263 @@
+    private @Nullable String commandUsername;
+    @SuppressWarnings("unused")
+    private @Nullable String commandPassword;
+    private long commandConnectTimeout = DEFAULT_CONNECT_TIMEOUT.toMillis();
+    private long commandRequestTimeout = DEFAULT_REQUEST_TIMEOUT.toMillis();
+    private String commandContentType = DEFAULT_CONTENT_TYPE;
+    @SuppressWarnings("unused")
+    private @Nullable String commandRequestTransform;
+    @SuppressWarnings("unused")
+    private @Nullable String commandResponseTransform;
+
+    public Optional<StateRequest> getStateRequest(final BundleContext bundleContext) throws IllegalArgumentException {
+        return Optional.ofNullable(this.stateUrl).map(stateUrl -> {
+            try {
+                final Optional<Transform> stateResponseTransform = Optional.ofNullable(this.stateResponseTransform)
+                        .map(s -> Transform.parse(bundleContext, s));
+                return new StateRequest(
+                        new URL(stateUrl),
+                        Optional.ofNullable(stateUsername),
+                        Optional.ofNullable(statePassword),
+                        Duration.ofMillis(this.stateConnectTimeout),
+                        Duration.ofMillis(this.stateRequestTimeout),
+                        Duration.ofMillis(this.stateRefreshInterval),
+                        stateResponseTransform
+                );
+            } catch (final MalformedURLException e) {
+                throw new IllegalArgumentException("Invalid stateUrl: " + e.getMessage());
+            } catch (final IllegalArgumentException e) {
+                throw new IllegalArgumentException("Invalid stateResponseTransform: " + e.getMessage(), e);
+            }
+        });
+    }
+
+    public Optional<CommandRequest> getCommandRequest(final BundleContext bundleContext) throws IllegalArgumentException {
+        return Optional.ofNullable(this.commandUrl).map(commandUrl -> {
+            final Method commandMethod;
+            try {
+                commandMethod = Method.valueOf(this.commandMethod);
+            } catch (final IllegalArgumentException e) {
+                throw new IllegalArgumentException("Invalid commandMethod", e);
+            }
+            final Optional<Transform> commandRequestTransform;
+            try {
+                commandRequestTransform = Optional.ofNullable(this.commandRequestTransform)
+                        .map(s -> Transform.parse(bundleContext, s));
+            } catch (final IllegalArgumentException e) {
+                throw new IllegalArgumentException("Invalid commandRequestTransform: " + e.getMessage());
+            }
+            final Optional<Transform> commandResponseTransform;
+            try {
+                commandResponseTransform = Optional.ofNullable(this.commandResponseTransform)
+                        .map(s -> Transform.parse(bundleContext, s));
+            } catch (final IllegalArgumentException e) {
```

Comment:
This IMHO now how exception handling should be used in Java, every line of code in this method has its own exception handling. The idea was that you could group and that it does not mix with you logic flow.

Reply:
Okay, let's take one step back. 

I think we can safely stop throwing `IllegalArgumentException` (3x in Transform && 6x in this class), and create specific ones there. 

And if we can catch one then throw back something else.

It could then look like:

```java
public Optional<CommandRequest> getCommandRequest(final BundleContext bundleContext) throws IllegalArgumentException {
    return Optional.ofNullable(this.commandUrl).map(commandUrl -> {
        Optional<Transform> commandRequestTransform = Optional.ofNullable(this.commandRequestTransform)
                .map(s -> Transform.parse(bundleContext, s));

        Optional<Transform> commandResponseTransform = Optional.ofNullable(this.commandResponseTransform)
                .map(s -> Transform.parse(bundleContext, s));

        try {
            return new CommandRequest(
                    getCommandMethod(),
                    new URL(commandUrl),
                    Optional.ofNullable(commandUsername),
                    Optional.ofNullable(commandPassword),
                    Duration.ofMillis(this.commandConnectTimeout),
                    Duration.ofMillis(this.commandRequestTimeout),
                    this.commandContentType,
                    commandRequestTransform,
                    commandResponseTransform
            );
        } catch (MalformedURLException e) {
            throw new InvalidConfigurationException(String.format("Invalid value '%s' for commandUrl  %s", commandUrl, e.getMessage()));
        }
    });
}

private Method getCommandMethod() {
    try {
        return Method.valueOf(this.commandMethod);
    } catch (IllegalArgumentException e) {
        throw new InvalidConfigurationException(String.format("Invalid value '%s' for commandMethod", this.commandMethod), e);
    }
}
```

I think the source of this problem is that we have a config which is not in the enum types and not yet validated and then immediately go from the 'broken' model to some request. 

Maybe it would even make more sense to separate validation of the model from actual construction of actions. That is also how other bindings tend to do this. 

Because we should also provide feedback about this incorrect config to the user via the Thing-status, not just through an exception in a log.

Reply:
I like your new suggestion, but it still causes the loss of one bit of information: if one of the transformations specified is invalid, this doesn't let us tell the user which one.

I just pushed a change that more or less does what you're asking; let me know what you think.

> Maybe it would even make more sense to separate validation of the model from actual construction of actions. That is also how other bindings tend to do this.

Is there a binding that you think does this particularly well that I could use as an example?

> Because we should also provide feedback about this incorrect config to the user via the Thing-status, not just through an exception in a log.

Sure, that's exactly why I've done it this way: these methods get called in the thing handler's `initialize()`, which catches the exception and updates the Thing status appropriately with a detailed message describing which configuration entry was invalid and why.

Reply:
The configuration class is capable of parsing textual values and assign the corresponding enum value to a field, and will throw on invalid user input.

Reply:
Ah, great, can at least eliminate the need to parse the HTTP method manually.